librz/analysis/arch/ppc/ppc_il.c

```diff
@@ -0,0 +1,882 @@
+ */
+RZ_IPI const char *ppc_get_cr_name(const ut8 x) {
+	switch (x) {
+	default:
+		RZ_LOG_WARN("Cannot return cr%" PFMT32d ". There exists no such register.\n", x);
+		return NULL;
+	case 0:
+		return "cr0";
+	case 1:
+		return "cr1";
+	case 2:
+		return "cr2";
+	case 3:
+		return "cr3";
+	case 4:
+		return "cr4";
+	case 5:
+		return "cr5";
+	case 6:
+		return "cr6";
+	case 7:
+		return "cr7";
+	}
+}
+
+/**
+ * \brief Synchronizes the CR register with the CR0-CR7 registers.
+ * Since CR contains CR0-CR7 but are separated in the register profile
+ * this function should be called before CR as a whole is read and after it was written.
+ *
+ * \param to_cr True: CR0-CR7 are copied to CR. False: CR is copied to CR0-CR7 according to \p cr_mask.
+ * \param cr_mask Masks the bits which are copied from CR to CR0-CR7. Ignored if \p crx_to_cr == true.
+ * \return RzILOpEffect* Sequence of effects to sync the CR/CRx registers.
+ */
+RZ_IPI RZ_OWN RzILOpEffect *ppc_sync_crx_cr(const bool crx_to_cr, const ut32 cr_mask) {
+	RzILOpEffect *sync;
+	if (crx_to_cr) {
+		sync = SETL("cr", LOGOR(UNSIGNED(32, VARG("cr7")), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr6")), U8(0x4)), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr5")), U8(0x8)), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr4")), U8(0xc)), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr3")), U8(0x10)), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr2")), U8(0x14)), LOGOR(SHIFTL0(UNSIGNED(32, VARG("cr1")), U8(0x18)), SHIFTL0(UNSIGNED(32, VARG("cr0")), U8(0x1c))))))))));
+		return sync;
+	}
+	sync = SEQN(10,
+		SETL("cr_mask", U32(cr_mask)),
+		SETL("crm", LOGAND(VARL("cr"), U32(cr_mask))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf))), EMPTY(), SETG("cr7", UNSIGNED(4, VARL("crm")))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf0))), EMPTY(), SETG("cr6", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x4))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf00))), EMPTY(), SETG("cr5", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x8))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf000))), EMPTY(), SETG("cr4", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0xc))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf0000))), EMPTY(), SETG("cr3", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x10))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf00000))), EMPTY(), SETG("cr2", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x14))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf000000))), EMPTY(), SETG("cr1", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x18))))),
+		BRANCH(IS_ZERO(LOGAND(VARL("cr_mask"), U32(0xf0000000))), EMPTY(), SETG("cr0", UNSIGNED(4, SHIFTR0(VARL("crm"), U8(0x1c))))));
+	return sync;
+}
+
+/**
+ * \brief Returns the mask for a given fxm operand.
+ * For details look up the "mtcrf" instruction in the Power ISA
+ *
+ * \param fmx The fmx value.
+ * \return ut32 The mask for writing to the CR register.
+ */
+RZ_IPI ut32 ppc_fmx_to_mask(const ut8 fmx) {
+	ut32 x = 0xf;
+	return (
+		(fmx & 0x80 ? x << 0x1c : 0) |
```

Comment:
In general, for shifter values, it's better to use decimal numbers rather than hexadecimal.